plugins/tabular_dataset.cc

```diff
@@ -349,6 +353,57 @@ struct TabularDataset::TabularDataStore: public ColumnIndex, public MatrixView {
                 columnHashes.push_back(ch);
             }
     }
+
+    void commit()
+    {
+        if (!mutable_chunks.empty())
+        {
+            mutable_chunks[0].freeze();
+            finalize(mutable_chunks, mutable_chunks[0].rowCount());
+            mutable_chunks.resize(0);
+        }
+    }
+
+    void recordRow(const RowName & rowName, const std::vector<std::tuple<ColumnName, CellValue, Date> > & vals)
+    {
+        if (rowCount > 0)
+            HttpReturnException(400, "Tabular dataset has already been committed, cannot add more rows");
+
+        if (mutable_chunks.empty())
+        {
+            //need to create the mutable chunk
+            vector<ColumnName> columnNames;
+            //The first recorded row will determine the columns
+            ML::Lightweight_Hash<ColumnHash, int> inputColumnIndex;
+            for (unsigned i = 0;  i < vals.size();  ++i) {
+                const ColumnName & c = std::get<0>(vals[i]);
+                ColumnHash ch(c);
+                if (!inputColumnIndex.insert(make_pair(ch, i)).second)
+                    throw HttpReturnException(400, "Duplicate column name in tabular dataset entry",
+                                              "columnName", c.toString());
+                columnNames.push_back(c);
+            }
+
+            initialize(columnNames, inputColumnIndex);
+
+            mutable_chunks.emplace_back(columnNames.size(), TABULAR_DATASET_DEFAULT_ROWS_PER_CHUNK);
+        }
+
+        std::vector<CellValue> orderedVals(columnNames.size());
+        Date ts;
+        for (unsigned i = 0;  i < vals.size();  ++i) {
+            const ColumnName & c = std::get<0>(vals[i]);
+            auto iter = columnIndex.find(c);
+            if (iter == columnIndex.end())
+                throw HttpReturnException(400, "New column name while recording tow in tabular dataset", "columnName", c.toString());
+
+            orderedVals[iter->second] = std::get<1>(vals[i]);
+
+            ts = (i == 0) ? std::get<2>(vals[i]) : std::max(ts, std::get<2>(vals[i]));
+        }
+
+        mutable_chunks[0].add(rowName, ts, orderedVals.data());
```

Comment:
Is it true that for tabular dataset each column has the same timestamp for a given row?  Isn't it this too limiting?